include/ck/tensor_operation/gpu/device/device_layernorm.hpp

```diff
@@ -51,27 +86,74 @@ struct DeviceLayernorm : public BaseOperator
 
     using PassThrough = tensor_operation::element_wise::PassThrough;
 
-    // Used for freeloading of some handy functions from DeviceReduceMultiBlock
-    using Reduction = DeviceReduceMultiBlock<XDataType,
-                                             AccDataType,
-                                             YDataType,
-                                             Rank,
-                                             NumReduceDim,
-                                             reduce::Add,
-                                             PassThrough,             // InElementwiseOperation
-                                             AccElementwiseOperation, // AccElementwiseOperation
-                                             InMemoryDataOperationEnum::Set,
-                                             false, // PropagateNan
-                                             false, // OutputIndex
-                                             false, // HaveIndexInputIfOutputIndex
-                                             BlockSize,
-                                             MThreadClusterSize,
-                                             KThreadClusterSize,
-                                             MThreadSliceSize,
-                                             KThreadSliceSize,
-                                             XYSrcVectorDim,
-                                             XSrcVectorSize,
-                                             1>; // YDstVectorSize
+    static constexpr index_t M_BlockTileSize = MThreadClusterSize * MThreadSliceSize;
+    static constexpr index_t K_BlockTileSize = KThreadClusterSize * KThreadSliceSize;
+
+    static auto MakeSrc2dDescriptor(const std::vector<index_t>& inLengths,
+                                    const std::vector<index_t>& inStrides,
+                                    int blkGroupSize,
+                                    int numBlockTileIteration)
+    {
+        constexpr index_t NumInvariantDim  = Rank - NumReduceDim;
+        static constexpr index_t numSrcDim = Rank;
+        static constexpr bool reduceAllDim = (NumInvariantDim == 0);
+
+        const auto tupleSrcLengths = make_tuple_from_array(inLengths, Number<numSrcDim>{});
+        const auto tupleSrcStrides = make_tuple_from_array(inStrides, Number<numSrcDim>{});
+
+        const auto inDesc = make_naive_tensor_descriptor(tupleSrcLengths, tupleSrcStrides);
+
+        const auto in_grid_desc_m_k = [&]() {
+            if constexpr(reduceAllDim)
```

Comment:
For LayerNorm,  you can always assume  `reduceAllDim` is false.    You can add `static_assert(Rank > NumReduceDim, "xxx")`

Reply:
No, if the batchsize == 0, which mean user apply layernorm on entire x, it is reduceAllDim .
In normal case of NCHW, user will reduce CHW.
However, user still can pass CHW tensor if user pass data without batch.

Reply:
I see